### PR TITLE
Add support for -t/--template option to export

### PR DESCRIPTION
**Pull request by msabramo**
This is a honcho analogue of the corresponding options to `foreman export` -- See:

- http://ddollar.github.io/foreman/#EXPORTING
- https://github.com/ddollar/foreman/pull/46

This lets you specify a custom Jinja template to be used when exporting to a supervisord config file.

Fixes: GH-88

See: https://github.com/nickstenning/honcho/issues/88

OK, I just updated this so that `--template` is a directory. I'm not quite sure how to test this in an automated way, but I did do manual testing.

```
[marca@marca-mac2 export_test]$ ls -l upstart_template.d
total 24
-rw-r--r--+ 1 marca  staff  353 Feb 23 06:01 master.conf
-rw-r--r--+ 1 marca  staff  380 Feb 23 06:02 process.conf
-rw-r--r--+ 1 marca  staff  119 Feb 23 06:02 process_master.conf

[marca@marca-mac2 export_test]$ honcho export upstart upstart-export --template upstart_template.d
2015-02-23 07:59:43 [20837] [INFO] Writing 'upstart-export/export_test.conf'
2015-02-23 07:59:43 [20837] [INFO] Writing 'upstart-export/export_test-ansvc.conf'
2015-02-23 07:59:43 [20837] [INFO] Writing 'upstart-export/export_test-ansvc-1.conf'

[marca@marca-mac2 export_test]$ honcho export upstart upstart-export
2015-02-23 08:00:16 [20841] [INFO] Writing 'upstart-export/export_test.conf'
2015-02-23 08:00:16 [20841] [INFO] Writing 'upstart-export/export_test-ansvc.conf'
2015-02-23 08:00:16 [20841] [INFO] Writing 'upstart-export/export_test-ansvc-1.conf'

[marca@marca-mac2 export_test]$ ls -l supervisor-template.d
total 8
-rw-r--r--+ 1 marca  staff  776 Dec 23 08:09 supervisord.conf

[marca@marca-mac2 export_test]$ honcho export supervisord out --template=supervisor-template.d
2015-02-23 08:00:39 [20847] [INFO] Writing 'out/export_test.conf'

[marca@marca-mac2 export_test]$ honcho export supervisord out
2015-02-23 08:01:05 [20855] [INFO] Writing 'out/export_test.conf'
```

### CODE REVIEW

**Comment by msabramo** [comment]
OK, I added a couple of simple unit tests in 6a5a97f.

**Comment by msabramo** [comment]
@nickstenning, @slafs: Please review when you get a chance.

**Comment by msabramo** [comment]
Hey, sorry to bother, I wonder if I can do anything to move this along, as we are planning a project at work where we want to eliminate dozens of manually-curated supervisor config files and replace them with `Procfile`s and a process that automatically generates the supervisor configs.

I will probably need this for that task.

**Comment by slafs** [comment]
Hi @msabramo! I'm little busy right now. Will take a look at this on weekend if I have some time.

**Comment by msabramo** [comment]
@nickstenning: Do you have time to look at this? If not, that's fine -- though I may need to devise a solution for generating supervisor configs that doesn't use honcho and I'd rather use honcho if possible, as it works nicely and looks like foreman so it's more familiar to some people, etc.

File: doc/export.rst [hunk 1]
```diff
@@ -32,6 +32,9 @@ Options
     -l DIR, --log DIR
         Specify the directory to place process logs in.
 
+    -t TEMPLATE, --template TEMPLATE
+        Specify alternate template to use for creating export files.
```
Comment (nickstenning) [inline doc/export.rst]: I think this should be `template directory`.

File: honcho/command.py [hunk 2]
```diff
@@ -135,6 +136,10 @@ def command_export(args):
     help="the shell that should run the application",
     default='/bin/sh', type=str)
 parser_export.add_argument(
+    '-t', '--template',
+    help="alternate template to use for creating export files",
```
Comment (nickstenning) [inline honcho/command.py]: Ditto: `s/template/template dir/`

File: honcho/export/base.py [hunk 3]
```diff
@@ -19,17 +20,22 @@
...
+        self._template_dir = template_dir or self.default_template_dir
 
-    def get_template(self, path):
+    def get_template(self, name):
         """
         Retrieve the template at the specified path. Returns an instance of
         :py:class:`Jinja2.Template` by default, but may be overridden by
         subclasses.
         """
+        path = os.path.join(self._template_dir, name)
```
Comment (nickstenning) [inline honcho/export/base.py]: Perhaps don't do this here, and just change the search path of the `FileSystemLoader` below?

File: honcho/export/base.py [hunk 4]
```diff
@@ -19,17 +20,22 @@
...
 class BaseExport(object):
-    def __init__(self, template_env=None):
+    # Override this in subclasses -- e.g.: `supervisord`, `upstart`, etc.
+    default_template_dir = ''
+
+    def __init__(self, template_dir=None, template_env=None):
         if template_env is None:
             template_env = _default_template_env()
         self._template_env = template_env
+        self._template_dir = template_dir or self.default_template_dir
```
Comment (nickstenning) [inline honcho/export/base.py]: I think I'd rather you made the `_template_dir` option `None` by default, and only changed the `jinja2` loader to a `FileSystemLoader` when `self._template_dir is not None`.

File: honcho/test/fixtures/custom_supervisord.conf [hunk 5]
```diff
@@ -0,0 +1,22 @@
+# Generated from custom template: honcho/test/fixtures/custom_supervisord.conf
```
Comment (nickstenning) [inline honcho/test/fixtures/custom_supervisord.conf]: I think having a simpler template just to test the `-t` functionality would be fine.

File: doc/export.rst (hunk 1 above)
Comment (msabramo) [inline doc/export.rst]: Done. Thanks!

File: honcho/command.py (hunk 2 above)
Comment (msabramo) [inline honcho/command.py]: Done. Thanks!

File: honcho/test/fixtures/custom_supervisord.conf (hunk 5 above)
Comment (msabramo) [inline honcho/test/fixtures/custom_supervisord.conf]: Done. Thanks!

File: honcho/export/base.py (hunk 4 above)
Comment (msabramo) [inline honcho/export/base.py]: Good idea! See new version.

File: honcho/export/base.py (hunk 3 above)
Comment (msabramo) [inline honcho/export/base.py]: Good idea! See new version.

**Comment by msabramo** [comment]
@nickstenning: Thanks for the review! Addressed your comments and I do think it's easier to understand now. Let me know what you think.

**Comment by nickstenning** [comment]
Great! I've cleaned up a couple of bits and pieces and merged this as 5886e1849cd6690883b3a08b2143d2e4e52cb497. Thanks Marc!

**Comment by msabramo** [comment]
Thanks you too for the great feedback and for open sourcing honcho in the first place!

**Comment by msabramo** [comment]
Feature available in Honcho 0.6.5: https://pypi.python.org/pypi/honcho/0.6.5

**Pull request closed without merging**